### PR TITLE
redis dataset implementation

### LINKED ISSUE #966: Redis.PickleDataset
## Description
A dataset for reading and writing data to a Redis cache. 

## Context
Redis is a super fast key value store. This would allow users to retrieve data shared between users in a super fast way.

## Possible Implementation
Currently, I use redis with pickle dumps and pickle loads as mentioned here: https://stackoverflow.com/a/20400288/12403925. I currently have my own custom redis.PickleDataSet that I could contribute.

## Possible Alternatives
Other possibilities include saving a JSON to Redis but pickle would allow saving any python object. Pickle is also much faster.
Furthermore, the pyarrow serialization/ deserialization methods have been deprecated so we cannot use it anymore.
https://stackoverflow.com/questions/57949871/how-to-set-get-pandas-dataframes-into-redis-using-pyarrow

### PR DESCRIPTION
**Pull request by juan-carlos-calvo**
## Description

Resolves #966 

## Development notes

Just added support for kwargs in `redis.StrictRedis.from_url` and `redis.StrictRedis.set` to the implementation in #966 . Also added a unit test.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [x] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1176"></a>

### CODE REVIEW

File: kedro/extras/datasets/redis/redis_dataset.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        return dict(key=self._key, **self._from_url_args)
+
+    # `redis_db` mypy does not work since it is optional and optional is not
+    # accepted by pickle.loads.
+    def _load(self) -> Any:
+        return pickle.loads(self._redis_db.get(self._key))  # type: ignore
+
+    def _save(self, data: Any) -> None:
+        try:
+            self._redis_db.set(self._key, pickle.dumps(data), **self._save_args)
```
Comment (datajoely) [inline kedro/extras/datasets/redis/redis_dataset.py:96]: Do we want to support the other pickling libraries like Joblib etc? Perhaps you could wrap `PickleDataSet`.

**Comment by datajoely** [comment]
Amazing work - please remember to add references to the sphinx documentation and JSON schema definition :)

File: kedro/extras/datasets/redis/redis_dataset.py (hunk 1 above)
Comment (juan-carlos-calvo) [reply kedro/extras/datasets/redis/redis_dataset.py:96]: yeah, we probably do. Will include

**Comment by datajoely** [comment]
cc @lucasjamar

File: kedro/extras/datasets/redis/redis_dataset.py (hunk 1 above)
Comment (juan-carlos-calvo) [reply kedro/extras/datasets/redis/redis_dataset.py:96]: not sure if wrapping pickle dataset will work though as pickle dataset uses `.dump` (to a `fs_file`) whereas here we just use `.dumps` to obtain just the serialized string.

**Comment by lucasjamar** [comment]
Thx a lot @juan-carlos-calvo . I didnt find the time to do thid but i appreciate that you toke it over :)

**Comment by juan-carlos-calvo** [comment]
I've added the suggested changes @lucasjamar @datajoely

**Comment by juan-carlos-calvo** [comment]
> Amazing work - I'll wait for the core team to review, the only last suggestion I have is to question whether `redis.PickleDataSet` is a more accurate description?

good point. Renamed.

File: kedro/extras/datasets/redis/redis_dataset.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
+"""``PickleDataSet`` loads/saves data from/to a Redis database. The underlying
+functionality is supported by the redis library, so it supports all allowed
+options for instantiating the redis app ``from_url`` and setting a value."""
+
+import importlib
+import os
+import pickle
```
Comment (MerelTheisenQB) [inline kedro/extras/datasets/redis/redis_dataset.py:7]: Is this import needed?

File: kedro/extras/datasets/redis/redis_dataset.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+        if not (
+            hasattr(imported_backend, "loads") and hasattr(imported_backend, "dumps")
+        ):
+            raise ValueError(
+                f"Selected backend '{backend}' should satisfy the pickle interface. "
+                "Missing one of `loads` and `dumps` on the backend."
+            )
+
+        self._backend = imported_backend
```
Comment (MerelTheisenQB) [inline kedro/extras/datasets/redis/redis_dataset.py:132]: @AntonyMilneQB [created a PR to make sure the `pickle.PickleDataset` is deepcopy-able](https://github.com/kedro-org/kedro/pull/1213). Do we want this dataset to be deepcopy-able as well? If not, I think it's worth adding a note that this dataset doesn't work with the `ParallelRunner`.

File: tests/extras/datasets/redis/test_redis_dataset.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Copyright (c) 2016 - present
+# QuantumBlack Visual Analytics Ltd (a McKinsey company).
+# All rights reserved.
+#
+# This software framework contains the confidential and proprietary information
+# of QuantumBlack, its affiliates, and its licensors. Your use of these
```
Comment (MerelTheisenQB) [inline tests/extras/datasets/redis/test_redis_dataset.py:6]: We don't want these headers anymore now Kedro is part of the Linux Foundation.

File: tests/extras/datasets/redis/test_redis_dataset.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    io = PickleDataSet(redis_key, from_url_args, save_args)
+    io.save(dummy_object)
+    loaded_dummy_object = io.load()
+    from_url_mocker.assert_called_once_with(**from_url_args)
+    set_mocker.assert_called_once_with(redis_key, serialized_dummy_object, **save_args)
+    get_mocker.assert_called_once_with(redis_key)
+    pd.testing.assert_frame_equal(
+        dummy_object,
+        loaded_dummy_object,
+    )
```
Comment (MerelTheisenQB) [inline tests/extras/datasets/redis/test_redis_dataset.py:66]: Could you add some more tests to check all dataset functionality works? Just look at any of our Kedro datasets as an example.

**Review by MerelTheisenQB — COMMENTED** [review]
Thanks for the contribution @juan-carlos-calvo

I've added some comments and questions. 

A new thing we do need to ask all our users to do now we're part of the Linux Foundation is to sign-off their commits - this is what the DCO plugin checks. Can you please try the following commands:

```
git rebase HEAD~5 --signoff
git push --force-with-lease origin feat/redis_dataset
```

This will amend all your commits with a sign-off, and push --force-with-lease will overwrite the one already pushed to the remote but it will not overwrite any work on the remote branch if more commits were added to the remote branch (by another team-member or coworker or what have you). It ensures you do not overwrite someone elses work by force pushing.

If you wish to set up your env to be more comfortable for future Kedro contributions, just run make sign-off to make sure all your commits are automatically signed off.

**Comment by lorenabalan** [comment]
Hi @juan-carlos-calvo , just checking in if you're still keen to complete this work? Happy to review & approve once Merel's comments have been addressed. 😊

**Comment by juan-carlos-calvo** [comment]
> 



> Hi @juan-carlos-calvo , just checking in if you're still keen to complete this work? Happy to review & approve once Merel's comments have been addressed. 😊

Hi @lorenabalan yes, I'm picking this up again. Today being my last day at the firm I'm handing in my laptop but will try to continue with the unit tests from gitpod within this coming days.

**Comment by datajoely** [comment]
Thanks @juan-carlos-calvo ! Forever a Kedroid :)

**Pull request closed without merging**